src/base/emu-i386/simx86/interp.c

```diff
@@ -890,9 +890,28 @@ intop3b:		{ int op = ArOpsFR[D_MO(opc)];
 /*6a*/	case PUSHbi:
 			Gen(O_PUSHI, mode, (signed char)Fetch(PC+1)); PC+=2; break;
 /*60*/	case PUSHA:
-			Gen(O_PUSHA, mode); PC++; break;
+			/* push order: eax ecx edx ebx esp ebp esi edi */
+			Gen(O_PUSH1, mode);
+			Gen(O_PUSH2, mode, Ofs_EAX);
+			Gen(O_PUSH2, mode, Ofs_ECX);
+			Gen(O_PUSH2, mode, Ofs_EDX);
+			Gen(O_PUSH2, mode, Ofs_EBX);
+			Gen(O_PUSH2, mode, Ofs_ESP);
+			Gen(O_PUSH2, mode, Ofs_EBP);
+			Gen(O_PUSH2, mode, Ofs_ESI);
+			Gen(O_PUSH2, mode, Ofs_EDI);
+			Gen(O_PUSH3, mode); PC++; break;
 /*61*/	case POPA:
-			Gen(O_POPA, mode); PC++; break;
+			Gen(O_POP1, mode);
+			Gen(O_POP2, mode, Ofs_EDI);
+			Gen(O_POP2, mode, Ofs_ESI);
+			Gen(O_POP2, mode, Ofs_EBP);
+			Gen(O_POP2, mode, Ofs_ESP); // overwritten in O_POP3
```

Comment:
Maybe replace with increment then?

Reply:
In that case I'd need a new flag, a bit like MPOPRM. But it doesn't matter that TheCPU.esp temporarily has an incorrect value as O_POP3 will set it correctly. Once can also write to TheCPU.ebx (which is then replaced with the correct value in the next O_POP2) if you feel that's safer...